File: open_raccoon/apps/raccoon_chat/lib/raccoon_chat/conversation_member.ex
defmodule RaccoonChat.ConversationMember do
  @moduledoc """
  Conversation membership schema with roles.
  """
  use Ecto.Schema
  import Ecto.Changeset

  @primary_key {:id, :binary_id, autogenerate: true}
  @foreign_key_type :binary_id
  schema "conversation_members" do
    belongs_to(:conversation, RaccoonChat.Conversation)
    belongs_to(:user, RaccoonAccounts.User)

    field(:role, Ecto.Enum, values: [:owner, :admin, :member], default: :member)
    field(:muted, :boolean, default: false)
    field(:last_read_at, :utc_datetime_usec)
    field(:joined_at, :utc_datetime_usec)
  end

  def changeset(member, attrs) do
    member
    |> cast(attrs, [:conversation_id, :user_id, :role, :muted, :last_read_at, :joined_at])
    |> validate_required([:conversation_id, :user_id])
    |> validate_inclusion(:role, [:owner, :admin, :member])
    |> unique_constraint([:conversation_id, :user_id])
    |> foreign_key_constraint(:conversation_id)
    |> foreign_key_constraint(:user_id)
  end
end

File: WaiAgents/Sources/WaiAgentsKit/Models/ConversationMember.swift
import Foundation

public struct ConversationMember: Codable, Identifiable, Sendable, Equatable {
    public let id: String
    public let conversationID: String
    public let userID: String
    public var role: MemberRole
    public var muted: Bool
    public var lastReadAt: Date?
    public let joinedAt: Date

    public enum MemberRole: String, Codable, Sendable {
        case owner
        case admin
        case member
    }

    enum CodingKeys: String, CodingKey {
        case id
        case conversationID = "conversationId"
        case userID = "userId"
        case role
        case muted
        case lastReadAt
        case joinedAt
    }

    public init(
        id: String,
        conversationID: String,
        userID: String,
        role: MemberRole = .member,
        muted: Bool = false,
        lastReadAt: Date? = nil,
        joinedAt: Date = Date()
    ) {
        self.id = id
        self.conversationID = conversationID
        self.userID = userID
        self.role = role
        self.muted = muted
        self.lastReadAt = lastReadAt
        self.joinedAt = joinedAt
    }
}
